File: euler341/euler341.py
#!/usr/bin/env pypy

import sys

sys.setrecursionlimit(2000000)

sys.path[-1] = "../common"

import common

golomb2 = [0, 1, 2]

def golomb2__(n):
    while n >= len(golomb2):
        golomb2.append( golomb2[-1] + (len(golomb2) - 1)*golomb_(len(golomb2) - 1) )
        if n % 100 == 0:
            print float(golomb2[-1])/10**18
    return golomb2[n]

def golomb2_(n):
    # print "!!!", n
    
    while n >= golomb2[-1]:
        golomb2__(len(golomb2))
    
    # i = 0
    # while n >= golomb2__(i):
    #     i += 1
    # return i - 1

    left = 0
    right = len(golomb2) - 1
    while left < right:
        mid = (left + right) / 2
        val = golomb2__(mid)
        # print "!", n, left, right, mid, val
        if val == n:
            return mid
        elif val > n:
            right = mid
        else:
            left = mid + 1
    while n >= golomb2[left]:
        left -= 1
    return left - 1

def golomb__(n):
    res = 1
    while 1 != n:
        n -= golomb2_(n - 1)
        res += 1
    return res

M = 1000
cache = [0]*M

def golomb_(n):
    if n < M:
        if 0 == cache[n]:
            cache[n] = golomb__(n)
        return cache[n]
    return golomb__(n)

assert golomb_(1000) == 86
assert golomb_(1000000) == 6137

@common.memoize
def golomb(n):
    while 1 != n:
        return 1
    return 1 + golomb(n - golomb(golomb(n - 1)))

if False:
    for i in xrange(1, 100):
        print i, golomb(i), golomb(golomb(i))

    sys.exit(0)

if False:
    for i in xrange(1, 1000):
        print i, golomb_(i), golomb(i)
        assert golomb_(i) == golomb(i)

    sys.exit(0)

if True:
    print golomb2_(10**18)
    sys.exit(0)

N = 10**6
res = 0
for i in xrange(1, N):
    # if i % 1000 == 0:
    print i, golomb2_(i*i*i), golomb_(i*i*i)
    res += golomb_(i*i*i)
print res

File: euler341/euler341.cpp
#include <cstdio>

#include <vector>

using namespace std;

typedef long long int ui64;
typedef vector<ui64> TIntVector;

static const ui64 initGolomb2[] = {0, 1, 2};
TIntVector golomb2(initGolomb2, initGolomb2 + 3);

static const ui64 N = 1000;
static const ui64 N3 = N*N*N;
static const ui64 N18 = 1000000000000000000ULL;

ui64 golomb_(ui64 n);

ui64 golomb2__(ui64 n) {
    while (n >= golomb2.size()) {
        golomb2.push_back( golomb2.back() + (golomb2.size() - 1)*golomb_(golomb2.size() - 1) );
    }
    return golomb2[n];
}

ui64 golomb2_(ui64 n) {
    while (n >= golomb2.back()) {
        golomb2__(golomb2.size());
    }

    size_t left = 0;
    size_t right = golomb2.size() - 1;
    while (left < right) {
        const size_t mid = (left + right) / 2;
        const ui64 val = golomb2[mid];
        if (val == n)
            return mid;
        else if (val > n)
            right = mid;
        else
            left = mid + 1;
    }
    while (n >= golomb2[left])
        --left;
    return left - 1;
}

static const ui64 initCache[] = {0, 1, 2};
TIntVector cache(initCache, initCache + 3);

void FillCache() {
    static const ui64 CACHE_SIZE = 2000000000;
    cache.resize(CACHE_SIZE);
    cache[1] = 1;
    for (size_t i = 2; i < CACHE_SIZE; ++i)
        cache[i] = 1 + cache[i - cache[cache[i - 1]]];
}

ui64 golomb_(ui64 n) {
    ui64 res = 0;
    while (n >= cache.size()) {
        ui64 curGolomb = golomb2_(n - 1);
        ui64 limit = golomb2[curGolomb];
        ui64 d = (n - limit + curGolomb - 1)/curGolomb;
        // printf("%lld %lld\n", curGolomb, d);
        if (d > 0) {
            n -= d*curGolomb;
            res += d;
        } else {
            n -= curGolomb;
            ++res;
        }
    }
    return res + cache[n];
}

int main() {
    printf("%lld %lld", golomb_(1000), golomb_(1000000));
    FillCache();
    printf(" %lld", golomb2_(N18));
    {
        ui64 res = 0;
        for (ui64 i = 1; i < 1000; ++i) {
            res += golomb_(i*i*i);
        }
        printf(" %lld\n", res);
    }
    {
        ui64 res = 0;
        for (ui64 i = 1; i < 1000000; ++i) {
            if ((i % 100) == 0) {
                printf("%lld %.2lf %lld\n", i, 100.0*double(i)/1000000, res);
                fflush(stdout);
            }
            res += golomb_(i*i*i);
        }
        printf("%lld\n", res);
    }
    return 0;
}
